table/block_based/block_based_table_iterator.cc

```diff
@@ -558,4 +667,86 @@ void BlockBasedTableIterator::FindReadAheadSizeUpperBound() {
   block_prefetcher_.SetUpperBoundOffset(start_offset +
                                         total_bytes_till_upper_bound);
 }
+
+void BlockBasedTableIterator::BlockCacheLookupForReadAheadSize(
+    uint64_t offset, size_t readahead_size, size_t& updated_readahead_size) {
+  updated_readahead_size = readahead_size;
+
+  // readahead_cache_lookup_ can be set false after Seek, if after Seek or Next
+  // there is SeekForPrev or any other backward operation.
+  if (!readahead_cache_lookup_) {
+    return;
+  }
+
+  assert(!DoesContainBlockHandles());
+  assert(index_iter_->value().handle.offset() == offset);
+
+  // Error. current offset should be equal to what's requested for prefetching.
+  if (index_iter_->value().handle.offset() != offset) {
+    return;
+  }
+
+  size_t current_readahead_size = 0;
+  size_t footer = table_->get_rep()->footer.GetBlockTrailerSize();
+
+  // Add the current block to block_handles_.
+  {
+    BlockHandleInfo block_handle_info;
+    block_handle_info.index_val_ = index_iter_->value();
+    block_handles_.emplace_back(std::move(block_handle_info));
+  }
+
+  // Current block is included in length. Readahead should start from next
+  // block.
+  index_iter_->Next();
+  is_index_at_curr_block_ = false;
+
+  while (index_iter_->Valid()) {
```

Comment:
Also need to check the index_iter status. If status is not ok, how is it surfaced to the table iterator status?

Reply:
Oh! Yeah good point. Let me add that to the BlockHandleInfo.

Reply:
If there is a case when Status is Not OK but Valid is true, it'll fail in that case.

Reply:
Yes, if status is not ok, valid will be false. But we just need to make sure the non-ok status is surfaced.

> If the status is not ok we won't add to the block handle. So when index_iter_ points to that block, it will return the status as usual. I'm assuming that if status is not ok, Valid will be false for index_iter_

Ok. So if index_iter_ status is not ok, then we stop the index iteration and IsIndexAtCurr() will be true once we reach that block?

Reply:
Yes! Right now it's index_iter_.Valid(). So if it's false we don't add it to BlockHandleInfo or iterate over it and once IsIndexAtCurr is true it means it point to that block and will return NotOK status.